BIP 181, 182, 183: BIPs for Utreexo (#1923)
bitcoin/bips: bip-0181.md

accumulator. This allows for the construction of extremely lightweight nodes capable of performing
the same validation as a full node, without the need to store the entire UTXO set.

This BIP defines how the Utreexo accumulator works, defining the data structure and algorithms used to
maintain the accumulator, as well as how to generate and verify inclusion proofs for elements in the accumulator.
It does not define how the accumulator is used in the Bitcoin protocol, but rather provides a foundation for future
BIPs that will define how to integrate Utreexo into Bitcoin validation and P2P network protocol.

## License

This document is licensed under the BSD-3-Clause license.

## Preliminaries

An accumulator is a cryptographic data structure that allows for the compact representation of a set,
enabling efficient membership proofs without requiring storage of the entire set. In the context of Utreexo,
the accumulator tracks the current set of unspent transaction outputs (UTXOs).

The Utreexo accumulator is based on an append-only Merkle tree design introduced in [^1],
which provides logarithmic-sized inclusion proofs. Utreexo extends this design to support dynamic updates,

> specifically enabling deletions from the set—a requirement for tracking UTXO spends in Bitcoin.

Reviewer: Parsing through the linked paper it claimed that the accumulator defined there is sound and strong.

With the extension here to make that accumulator dynamic, I suppose it is still correct, sound and strong?
Perhaps link to some resource on where that was explicitly studied

> To accommodate this, Utreexo changes the storage requirement from the accumulator design in [^1] to $O(log_2(N))$,

Reviewer 2: Specifying the logarithm base is redundant in big O notation, as changing the base is equivalent to multiplying by a constant factor.
```suggestion
To accommodate this, Utreexo changes the storage requirement from the accumulator design in [^1] to $O(log(N))$,
```

Any integer number of elements ($N$) can be represented as a forest of such trees. On average, a set of N elements will require
approximately $\frac{log_2(N)}{2}$ trees. The number and sizes of trees are determined by the binary representation of $N$:
each 1-bit corresponds to a tree, and its position in the binary encoding determines the size of that tree.

For example, a forest with 5 elements (binary `0b101`) would consist of two trees: one with 4 elements (representing the 2nd bit)
and one with 1 element (representing the 0th bit). A forest with 8 elements (`0b1000`) would require only a single 8-element tree,
as 8 is a power of 2.

More generally, for any N, the number of trees equals the number of set bits (1s) in the binary representation of N.
The size of each tree corresponds to the power of two represented by the position of each set bit.
For example, the decimal number 21 (binary `0b10101`) contains three 1-bits, meaning three trees are needed in the forest:
a 16-element tree ($2^4$), a 4-element tree ($2^2$), and a 1-element tree ($2^0$), with gaps at the 8-element ($2^3$)
and 2-element ($2^1$) positions.

> Each of the hashes in the forest can be referred by an integer label. This labeling is a convention we find easiest

Reviewer 2: ```suggestion
Each of the hashes in the forest can be referred to by an integer label. This labeling is a convention we find easiest
```

to use but does not directly affect the design of the accumulator; other labelling systems could also work and be
translated to this one.

We label positions starting at `0` on the bottom left, incrementing as we traverse the bottom row from left to right,
and then continue on to higher rows. There may be gaps in the label numbers when moving up a row; the label
numbers are "padded out" to the next perfect tree that could encompass the entire forest.

For example, a forest with 8 leaves will have a single tree and positions will be labeled like this:

```
14
|---------------\
12              13
|-------\       |-------\
08      09      10      11
|---\   |---\   |---\   |---\
00  01  02  03  04  05  06  07
```

While a forest with 7 leaves will look like this:

```

|---------------\
12
|-------\       |-------\
08      09      10
|---\   |---\   |---\   |---\
00  01  02  03  04  05  06

> ```

Reviewer 3: Nit: maybe worth it to briefly mention the roots in this example are 06, 10 and 12

When adding another leaf to the accumulator when it's already allocated $2^N$ leaves will result in
the accumulator resizing to hold $2^{N+1}$ leaves. For example, when adding a leaf to the accumulator
state here:

```
14
|---------------\
12              13
|-------\       |-------\
08      09      10      11
|---\   |---\   |---\   |---\
00  01  02  03  04  05  06  07
```

The new accumulator will look like so:

```

|-------------------------------\
28
|---------------\               |---------------\
24              25
|-------\       |-------\       |-------\       |-------\
16      17      18      19
|---\   |---\   |---\   |---\   |---\   |---\   |---\   |---\
00  01  02  03  04  05  06  07  08
```

The new accumulator with all the positions:

```
30
|-------------------------------\
28                              29
|---------------\               |---------------\
24              25              26              27
|-------\       |-------\       |-------\       |-------\
16      17      18      19      20      21      22      23
|---\   |---\   |---\   |---\   |---\   |---\   |---\   |---\
00  01  02  03  04  05  06  07  08  09  10  11  12  13  14  15
```

# Definitions

> - `hash` refers to a vector of 32 byte arrays.

Reviewer 3: This seems wrong, it reads as a `Vec<[u8; 32]>` (a vector of hashes).

I think we should only say "32-byte array".

```suggestion
- `hash` refers to a 32-byte array.
```

- `[]hash` refers to a vector of `hash`.
- `acc` refers to the Utreexo accumulator state. An `acc` is comprised of:
  - `roots` refers to the roots of the Merkle Trees. Represented as `[]hash`.

>   - `numleaves` refers to the number of total leaves added to the accumulator. Represented as uint64.

Reviewer 3: ```suggestion
  - `numleaves` refers to the total number of leaves ever added to the accumulator. Represented as uint64.
```

> - `root` refers to the top `hash` in a tree in the `acc`.

Reviewer 3: ```suggestion
- `root` refers to the top `hash` of one Merkle tree in the `acc`.
```

- `proof` is an inclusion proof for elements in the accumulator. It's comprised of two fields:
  - `targets` are the positions of the elements being proven. Represented as a vector of uint64.

>   - `proof` are the hashes needed to hash the roots. Represented as a `[]hash`. `proof` MUST be in ascending order by the node positions.

Reviewer 3: ```suggestion
  - `proof` are the sibling hashes needed to compute the corresponding accumulator roots. Represented as a `[]hash`. The hashes in `proof` MUST be ordered by ascending node position.
```

    The proof is considered invalid otherwise.

# Specification

The hash function SHA512/256[^2] is used for the hash operations in the accumulator.

An Utreexo accumulator implementation MUST support these 3 operations: Addition, Verification, and Deletion.

## Utility Functions

The following utility functions are required for performing accumulator operations:

**parent_hash(left, right):** Returns the hash of the concatenation of two child hashes (`left` and `right`).
If either child is `nil`, the result is simply the non-`nil` child (treated as if the tree has a single child at that position).
if both children are `nil`, the result is `nil`.

Implementation:

```python
def parent_hash(left: bytes, right: bytes) -> bytes:
    if right is None and left is None: return None
    if left is None: return right
    if right is None: return left

    return sha512_256(left + right)
```

**treerows(numleaves):** Returns the minimum number of bits required to represent `numleaves - 1`. This corresponds to the height of the largest tree in the forest. Returns `0` if `numleaves` is `0`.

> The reason for taking the minimum number of bits required for `numleaves-1` and not `numleaves` is because when `numleaves` is a power of two, we'd get an off-by-one error.

Reviewer 2: It would be nice to have consistent spacing around the minus sign, there's both `numleaves - 1` and `numleaves-1`. Same goes for the equals sign below.

                parent_pos = parent(target, total_rows)
                next_targets.append(parent_pos)
                targets[i] = parent_pos
                i += 2  # skip the sibling
                continue

            # Sibling is a needed proof position
            proof_positions.append(sibling(target))
            parent_pos = parent(target, total_rows)
            next_targets.append(parent_pos)
            targets[i] = parent_pos
            i += 1

        targets.sort()

    return proof_positions
```

### CalculateRoots

Both the Verification and Deletion operations depend on the Calculate Roots function.

- Inputs:
  - `acc.numleaves`.
  - `[]hash` that are the hashes for the `proof.targets`.
  - `proof`.

The passed in `[]hash` and `proof.targets` should be in the same order. The element at index `i` in `[]hashes` should
be the hash for element at index `i` in `proof.targets`. Otherwise the returned roots will be invalid.

The calculate roots algorithm is defined as `CalculateRoots(numleaves, []hash, proof) -> calculated_roots`:

- Check if length of `proof.targets` is equal to the length of `[]hash`. Return early if they're not equal.

> - map `proof.targets` to their hash.

Reviewer 2: ```suggestion
- Map `proof.targets` to their hash.
```

- Sort `proof.targets`.
- Loop until `proof.targets` are empty:
  - Pop off the first target in `proof.targets`. Pop off the associated `hash` as well.
  - If the target is a root, we append the current position's `hash` to the calculated_roots vector and continue.
  - Check if the next target in `proof.targets` is the right sibling of the current target. If it is, grab its hash as the sibling hash. Otherwise the next hash in `proof.proof` is the sibling hash. Raise error if `proof.proof` is empty.
  - Figure out if the sibling hash is on the left or the right.
  - Apply *parent_hash* to the current position's `hash` and the sibling `hash` with regards to their positioning.
  - Calculate parent position.
  - Insert parent position into the sorted `proof.targets`.
  - Map parent hash to the parent position.
- Return calculated_roots

> The algorithm implemented in python:

Reviewer 2: ```suggestion
The algorithm implemented in Python:
```

    position_hashes = {}
    for i, target in enumerate(proof.targets):
        position_hashes[target] = None if dels is None else dels[i]

    calculated_roots = []
    sortedTargets = sorted(proof.targets)
    while sortedTargets:
        pos = sortedTargets.pop(0)
        cur_hash = position_hashes.pop(pos)

        if isroot(pos, numleaves, tree_rows(numleaves)):
            calculated_roots.append(cur_hash)
            continue

        parent_pos, p_hash = parent(pos, tree_rows(numleaves)), bytes
        if sortedTargets and right_sibling(pos) == sortedTargets[0]:
            sib_pos = sortedTargets.pop(0)
            p_hash = parent_hash(cur_hash, position_hashes.pop(sib_pos))
        else:
            proofhash = proof.proof.pop(0)
            p_hash = parent_hash(proofhash, cur_hash) if is_right_sibling(pos) else parent_hash(cur_hash, proofhash)

        position_hashes[parent_pos] = p_hash
        bisect.insort(sortedTargets, parent_pos)

    return calculated_roots
```

## Addition

Addition adds a leaf to the accumulator. The added leaves are able to be verified of their
existence with an inclusion proof.

Inputs:
  - `acc`.
  - `hash` to be added.

> The Addition algorithm Add(`acc`, `hash`) is defined as:

Reviewer 2: ```suggestion
The addition algorithm Add(`acc`, `hash`) is defined as:
```

- From row 0 to and **including** `treerows(acc.numleaves)`
  - Break if there's no root at this row.

>   - remove the last root from `acc.roots`.

Reviewer 2: ```suggestion
  - Remove the last root from `acc.roots`.
```

    - Calculate the parent hash of the removed root and the `hash` to be added using *parent_hash*.
  - Make the result from `parent_hash` the new `hash`.
- Increment `acc.numleaves` by 1.
- Append `hash` to `acc.roots`.

> The algorithm implemented in python:

Reviewer 2: ```suggestion
The algorithm implemented in Python:
```

```python
def add(self, hash: bytes):
    for row in range(tree_rows(self.numleaves)+1):
        if not root_present(self.numleaves, row): break
        root = self.roots.pop()
        hash = parent_hash(root, hash)

    self.roots.append(hash)
    self.numleaves += 1
```

## Verification

- Inputs:
  - The accumulator state.
  - `[]hash` that are the hashes for the `proof.targets`.
  - `proof`.

> The Verification algorithm `Verify(acc, []hash, proof) -> bool` is defined as:

Reviewer 2: ```suggestion
The verification algorithm `Verify(acc, []hash, proof) -> bool` is defined as:
```

- Raise error if length of `[]hash` differ from `proof.targets`.
- Get modified_roots from `CalculateRoots(acc.numleaves, []hash, Proof)`.
- Get `root_idxs` from `getrootidxs`.
- Raise error if the length of `modified_roots` and `root_idxs` do not match.
- Attempt to match roots in modified_roots with roots in `acc`. Raise error if we don't find all the roots in the modified_roots in `acc`.
- Return `true`.

> The algorithm implemented in python:

Reviewer 2: ```suggestion
The algorithm implemented in Python:
```

```python
def verify(self, dels: [bytes], proof: Proof) -> bool:
    if len(dels) != len(proof.targets):
        raise("len of dels and proof.targets differ")

    root_candidates = calculate_roots(self.numleaves, dels, proof)
    root_idxs = getrootidxs(self.numleaves, proof.targets)

    if len(root_candidates) != len(root_idxs):
        raise("length of calculated roots from the proof and expected root count differ")

    for i, idx in enumerate(root_idxs):
        if self.roots[idx] != root_candidates[i]:
            raise("calculated roots from the proof and matched roots differ")

    return true
```

## Deletion

Deletion removes leaves from the accumulator. The deletion algorithm takes in a `proof` but it does not
verify that the proof is valid. It assumes that the passed in proof has already passed verification.

- Inputs:
  - The accumulator state.
  - `proof`.

> The Deletion algorithm `Delete(acc, Proof) -> acc` is defined as:

Reviewer 2: ```suggestion
The deletion algorithm `Delete(acc, Proof) -> acc` is defined as:
```

- Get the modified indexes of the roots `root_idxes` from `getrootidxs`.
- Get modified_roots from `Calculate_Roots(acc.numleaves, []positions, Proof)`.
- Replace the matching indexes from the `root_idxes` in `acc.roots` with `modified_roots`.

> The algorithm implemented in python:

Reviewer 2: ```suggestion
The algorithm implemented in Python:
```
